python/lsst/ts/atmcssimulator/issued_command.py

```diff
@@ -0,0 +1,114 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__all__ = ["IssuedCommand"]
+
+import asyncio
+
+
+class IssuedCommand:
+    """asyncio futures to track the progress of an issued low-level controller
+    command.
+
+    Parameters
+    ----------
+    timestamp : float
+        The TAI unix timestamp [sec] at which the command was issued.
+    name : str
+        The name of the command.
+
+    Attributes
+    ----------
+    ack_or_noack_timestamp : float
+        The TAI unix timestamp [sec] at which an ACK or NOACK was received.
+    success_or_fail_timestamp : float
+        The TAI unix timestamp [sec] at which a SUCCESS or FAIL was received.
+    ack : `asyncio.Future`
+        Future which ends as follows:
+
+        * result = None when the command is acknowledged.
+        * exception if the command is rejected.
+    done : `asyncio.Future`
+        Future which ends as follows:
+
+        * result = None when the command has ended successfully.
+        * exception if the command has ended in an error.
+    """
+
+    def __init__(self, timestamp: float, name: str) -> None:
+        self.timestamp = timestamp
+        self.name = name
+        self.ack_or_noack_timestamp: float = 0.0
+        self.success_or_fail_timestamp: float = 0.0
+        self.ack: asyncio.Future = asyncio.Future()
+        self.done: asyncio.Future = asyncio.Future()
+
+    def set_ack(self, timestamp: float) -> None:
+        """Report a command as acknowledged, meaning accepted to be executed.
+
+        Parameters
+        ----------
+        timestamp : `float`
+            TAI unix time [sec] of when the command was acknowledged.
+        """
+        self.ack_or_noack_timestamp = timestamp
+        if not self.ack.done():
+            self.ack.set_result(None)
+
+    def set_noack(self, timestamp: float) -> None:
+        """Report a command as not acknowledged, meaning there was an issue
+        with the command format (e.g. the command name or one of more of its
+        parameters were wrong).
+
+        Parameters
+        ----------
+        timestamp : `float`
+            TAI unix time [sec] of when the command was not acknowledged.
+        """
+        self.ack_or_noack_timestamp = timestamp
+        if not self.done.done():
+            self.done.set_exception(
+                RuntimeError(f"Command {self.name} was not acknowledged.")
+            )
+
+    def set_success(self, timestamp: float) -> None:
+        """Report a command as successfully executed.
+
+        Parameters
+        ----------
+        timestamp : `float`
+            TAI unix time [sec] of when the command was successfully executed.
+        """
+        self.success_or_fail_timestamp = timestamp
+        if not self.done.done():
+            self.done.set_result(None)
```

Comment:
All commands require both an ack and a done. Good point that the command needs to be acknowledged first. I opted to implement this in the CSC so it can log a warning if a command is set to done or fail when it hasn't been acknowledged yet.

Reply:
I strongly encourage you to NOT require this for tracking commands. It doubles the bandwidth of replies for commands coming at 20 Hz. That adds more load to the CSC to handle the unnecessarily-doubled number of replies. Just send a single "done", instead.

Also, a command that is rejected with "noack" should definitely NOT also require a "failed". "noack" is final. There is no point to sending a followup.

Reply:
As discussed today via Slack, both `ack` and `done` are required. Indeed, a `noack` means that no `done` will follow.